contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2413,3 +2418,290 @@ sp_dropserver_internal(PG_FUNCTION_ARGS)
+					"SELECT t3.orig_name as SchemaName, t1.proname as FunctionName, "
+					"CASE "
+						"WHEN t1.provolatile = 'v' THEN 'volatile' "
+						"WHEN t1.provolatile = 's' THEN 'stable' "
+						"ELSE 'immutable' "
+					"END AS Volatility "
+					"from (SELECT (aclexplode(proacl)).*, proname, provolatile, pronamespace from pg_proc WHERE prokind = 'f') t1 "
+					"JOIN pg_namespace t2 ON t1.pronamespace = t2.oid "
+					"JOIN sys.babelfish_namespace_ext t3 ON t3.nspname = t2.nspname "
+					"where t1.grantee = %d AND t3.dbid = sys.db_id() ORDER BY t3.orig_name, t1.proname", user_id
+				);
+		}
+		else
+		{
+			query = psprintf(
+					"SELECT t3.orig_name as SchemaName, CAST('%s' as sys.varchar) as FunctionName, "
+					"CASE "
+						"WHEN provolatile = 'v' THEN 'volatile' "
+						"WHEN provolatile = 's' THEN 'stable' "
+						"ELSE 'immutable' "
+					"END AS Volatility from pg_proc t1 "
+					"JOIN pg_namespace t2 ON t1.pronamespace = t2.oid "
+					"JOIN sys.babelfish_namespace_ext t3 ON t3.nspname = t2.nspname "
+					"where t1.oid = %u", function_name, function_id
+				);
+		}
+
+		PG_TRY();
+		{	
+			char nulls = 0;
+			MemoryContext savedPortalCxt;
+			SPIPlanPtr plan;
+			Portal portal;
+			DestReceiver *receiver;
+
+			savedPortalCxt = PortalContext;
+			if (PortalContext == NULL)
+				PortalContext = MessageContext;
+			if ((rc = SPI_connect()) != SPI_OK_CONNECT)
+				elog(ERROR, "SPI_connect failed: %s", SPI_result_code_string(rc));
+			PortalContext = savedPortalCxt;
+
+			if ((plan = SPI_prepare(query, 0, NULL)) == NULL)
+				elog(ERROR, "SPI_prepare(\"%s\") failed", query);
+
+			if ((portal = SPI_cursor_open(NULL, plan, NULL, &nulls, true)) == NULL)
+				elog(ERROR, "SPI_cursor_open(\"%s\") failed", query);
+
+			/*
+			* According to specifictation, sp_babelfish_volatility returns a result-set.
+			* If there is no destination, it will send the result-set to client, which is not allowed behavior of PG procedures.
+			* To implement this behavior, we added a code to push the result.
+			*/
+			receiver = CreateDestReceiver(DestRemote);
+			SetRemoteDestReceiverParams(receiver, portal);
+
+			/* fetch the result and return the result-set */
+			PortalRun(portal, FETCH_ALL, true, true, receiver, receiver, NULL);
+
+			receiver->rDestroy(receiver);
+			SPI_cursor_close(portal);
+
+			if ((rc = SPI_finish()) != SPI_OK_FINISH)
+				elog(ERROR, "SPI_finish failed: %s", SPI_result_code_string(rc));
+		}
+		PG_CATCH();
+		{
+			SPI_finish();
+			PG_RE_THROW();
+		}
+		PG_END_TRY();	
+	}
+	else
+	{	
+		/* downcase identifier if needed */
+		volatility = downcase_identifier(volatility, strlen(volatility), false, false);
+	
+		if(strcmp(volatility, "volatile") && strcmp(volatility, "stable") && strcmp(volatility, "immutable"))
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("\"%s\" is not a valid volatility", volatility)));
+
+		query = psprintf("ALTER FUNCTION %s %s;", function_signature, volatility);
```

Comment:
can I execute any DDL here, what if my function name is "stable; drop table x;"?

Reply:
No, it is not possible. I have already tested for injection and added testcases for it